drivers/gpio/sn74hc595.c

```diff
@@ -0,0 +1,123 @@
+/* Copyright 2022
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "sn74hc595.h"
+#include "spi_master.h"
+#include "debug.h"
+#include "timer.h"
+
+#ifdef CONSOLE_ENABLE
+uint32_t timer_now;
+#endif
+
+// uint8_t *data_buffer;
+uint8_t pinStates[SN74HC595_LENGTH];
+typedef enum { Low = 0, High = 1 } state;
+
+/* get length */
+uint8_t sn74hc595_length(void) {
+    return SN74HC595_LENGTH;
+}
+/*
+    init first
+*/
+void sn74hc595_init(void) {
+    setPinOutputPushPull(SN74HC595_SPI_SLAVE_SELECT_PIN);
+    return spi_init();
+}
+
+/*
+    set pin number on sn74hc595
+*/
+void sn74hc595_setPin(pin_t pin, bool set) {
+    // set pin state like binary 1000010
+    for (int current = (SN74HC595_LENGTH - 1); current >= 0; current--) {
+        if (pin < 8) {
+            int mask           = 1 << pin;
+            pinStates[current] = ((pinStates[current] & ~mask) | set << pin);
+        }
+        pin -= 8;
+    }
+    // then send it using spi
+    spi_start(SN74HC595_SPI_SLAVE_SELECT_PIN, SN74HC595_SPI_LSBFIRST, SN74HC595_SPI_MODE, SN74HC595_SPI_CLOCK_DIVISOR);
+    writePinLow(SN74HC595_SPI_SLAVE_SELECT_PIN);
+    spi_transmit(pinStates, SN74HC595_LENGTH);
+    writePinHigh(SN74HC595_SPI_SLAVE_SELECT_PIN);
+    spi_stop();
+
+// pinstates debug
+#if defined(CONSOLE_ENABLE) && defined(SN74HC595_DEBUG)
+    if (timer_elapsed32(timer_now) >= 1000) {
+        uprintf("PinStates HEX: ");
```

Comment:
if you use dprintf here, it only prints if debug_enable is set to true. which can be toggled at runtime.

Reply:
i will fix it on next, also will fix other line on next.